sql/rowexec/dml_iters.go

```diff
@@ -516,29 +524,35 @@ func (a *accumulatorIter) Next(ctx *sql.Context) (r sql.Row, err error) {
 		default:
 		}
 		if err == io.EOF {
-			res := a.updateRowHandler.okResult() // TODO: Should add warnings here
-
 			// TODO: The information flow here is pretty gnarly. We
 			// set some session variables based on the result, and
 			// we actually use a session variable to set
 			// InsertID. This should be improved.
 
-			// By definition, ROW_COUNT() is equal to RowsAffected.
-			ctx.SetLastQueryInfo(sql.RowCount, int64(res.RowsAffected))
-
 			// UPDATE statements also set FoundRows to the number of rows that
 			// matched the WHERE clause, same as a SELECT.
 			if ma, ok := a.updateRowHandler.(matchingAccumulator); ok {
 				ctx.SetLastQueryInfo(sql.FoundRows, ma.RowsMatched())
 			}
 
 			newLastInsertId := ctx.Session.GetLastQueryInfo(sql.LastInsertId)
-			if newLastInsertId != -1 {
-				res.InsertID = uint64(newLastInsertId)
-			} else {
+			if newLastInsertId == -1 {
 				ctx.Session.SetLastQueryInfo(sql.LastInsertId, oldLastInsertId)
 			}
 
+			res := a.updateRowHandler.okResult() // TODO: Should add warnings here
+
+			// For some update accumulators, we don't accurately track the last insert ID in the handler and need to set
+			// it manually in the result by getting it from the session. This doesn't work correctly in all cases and needs
+			// to be fixed. See comment in buildRowUpdateAccumulator in rowexec/dml.go
+			switch a.updateRowHandler.(type) {
+			case *onDuplicateUpdateHandler, *replaceRowHandler:
```

Comment:
not sure i follow what's special about these

Reply:
I'll update the comment.

They're accumulators that can be put into place for insert statements